File: src/Subsystems/Vision.cpp
#include "Vision.h"
#include "../RobotMap.h"
#include <vector>
#include <networktables/NetworkTableInstance.h>

Vision::Vision() {
	cam0 = new cs::UsbCamera("cam0", 0);
	cam0->SetResolution(320, 240);
	cam0->SetBrightness(10);
	cam1 = new cs::UsbCamera("cam1", 1);
	cam1->SetResolution(320, 240);
	cam1->SetBrightness(10);
	cvsink0 = new cs::CvSink("cam0cv");
	cvsink0->SetSource(*cam0);
	cvsink0->SetEnabled(true);
	cvsink1 = new cs::CvSink("cam1cv");
	cvsink1->SetSource(*cam1);
	cvsink1->SetEnabled(true);
	CameraServer::GetInstance()->StartAutomaticCapture(*cam0);
	CameraServer::GetInstance()->StartAutomaticCapture(*cam1);
	server = CameraServer::GetInstance()->GetServer();
	SetCamera(0);
	table = nt::NetworkTableInstance::GetDefault().GetTable("GRIP/AllDemContours");
	this->Update();
}

void Vision::Update() {
	centerX = table->GetNumberArray("centerX", llvm::ArrayRef<double>());
	centerY = table->GetNumberArray("centerY", llvm::ArrayRef<double>());
	height = table->GetNumberArray("height", llvm::ArrayRef<double>());
	width = table->GetNumberArray("width", llvm::ArrayRef<double>());
}

double Vision::GetCentralValue() {
	this->Update();
	double theCenterX = 0;
	for(unsigned int i = 0; i < centerX.size(); i++) {
		theCenterX += centerX[i];
	}
	if(centerX.size() != 0) { theCenterX /= centerX.size(); }
	return theCenterX;
}

void Vision::SetCamera(int camera) {
	if(camera == 0) {
		server.SetSource(*cam0);
	} else {
		server.SetSource(*cam1);
	}
}

File: src/main/java/frc/team2410/robot/Subsystems/Vision.java
package frc.team2410.robot.Subsystems;

import edu.wpi.cscore.CvSink;
import edu.wpi.cscore.UsbCamera;
import edu.wpi.cscore.VideoSink;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.wpilibj.*;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.team2410.robot.RobotMap;

import static frc.team2410.robot.RobotMap.*;

public class Vision
{
	private NetworkTable table;
	private Number[] centerX;
	private Number[] centerY;
	private Number[] area;
	//private DigitalOutput light;
	private AnalogInput rangeFinder;
	
	public Vision() {
		table = NetworkTableInstance.getDefault().getTable("GRIP/AllDemContours");
		this.update();
		//light = new DigitalOutput(CAMERA_LIGHT);
		rangeFinder = new AnalogInput(RANGE_FINDER);
	}
	
	public double getDistanceAway() {
		return this.rangeFinder.getVoltage()/SONAR_VOLTS_PER_INCH;
	}
	
	private void update() {
		centerX = table.getEntry("centerX").getNumberArray(new Number[0]);
		centerY = table.getEntry("centerY").getNumberArray(new Number[0]);
		area = table.getEntry("area").getNumberArray(new Number[0]);
	}
	
	public double[] getCentralValue() {
		this.update();
		double x = 0;
		double y = 0;
		double greatestArea = 0;
		for(int i = 0; (i < area.length) && (i < centerX.length) && (i < centerY.length); i++) {
			if((double)area[i] > greatestArea) {
				x = (double)centerX[i];
				y = (double)centerY[i];
				greatestArea = (double)area[i];
			}
		}
		return new double[] {x, y};
	}
	/*public void setLight(boolean on) {
		light.set(on);
	}*/
}
